falcon/media/multipart.py

```diff
@@ -0,0 +1,402 @@
+        self._parse_options = parse_options
+
+    def __iter__(self):
+        delimiter = self._dash_boundary
+        stream = self._stream
+        max_headers_size = self._parse_options.max_body_part_headers_size
+        remaining_part_count = self._parse_options.max_body_part_count
+
+        while True:
+            # NOTE(vytas): Either exhaust the unused stream part, or skip
+            #   the prologue.
+            stream.pipe_until(delimiter)
+            stream.read(len(delimiter))
+
+            if not delimiter.startswith(CRLF):
+                # NOTE(vytas): RFC 7578, section 4.1.
+                #   As with other multipart types, the parts are delimited with
+                #   a boundary delimiter, constructed using CRLF, "--", and the
+                #   value of the "boundary" parameter.
+                delimiter = CRLF + delimiter
+
+            separator = stream.read_until(CRLF, 2, MultipartParseError)
+            if separator == b'--':
+                if stream.peek(2) != CRLF:
+                    raise MultipartParseError('unexpected form epilogue')
+                break
+            elif separator:
+                raise MultipartParseError('unexpected form structure')
+
+            headers = {}
+            headers_block = stream.read_until(_CRLF_CRLF, max_headers_size,
+                                              MultipartParseError)
+            stream.read(4)
+
+            for line in headers_block.split(CRLF):
+                name, sep, value = line.partition(b': ')
+                if sep:
+                    name = name.lower()
+
+                    # NOTE(vytas): RFC 7578, section 4.5.
+                    #   This use is deprecated for use in contexts that support
+                    #   binary data such as HTTP. Senders SHOULD NOT generate
+                    #   any parts with a Content-Transfer-Encoding header
+                    #   field.
+                    #
+                    #   Currently, no deployed implementations that send such
+                    #   bodies have been discovered.
+                    if name == b'content-transfer-encoding':
+                        raise MultipartParseError(
+                            'the deprecated Content-Transfer-Encoding header '
+                            'field is unsupported')
+                    # NOTE(vytas): RFC 7578, section 4.8.
+                    #   Other header fields MUST NOT be included and MUST be
+                    #   ignored.
+                    elif name in _ALLOWED_CONTENT_HEADERS:
+                        headers[name] = value
+
+            remaining_part_count -= 1
+            if remaining_part_count == 0:
+                raise MultipartParseError(
+                    'maximum number of form body parts exceeded')
+
+            yield BodyPart(stream.delimit(delimiter, MultipartParseError),
+                           headers, self._parse_options)
+
+        stream.exhaust()
```

Comment:
I need to go check this, but I remember elsewhere seeing:

```py
if handler.exhaust_stream:
    self.bounded_stream.exhaust()
```

Are we potentially disregarding the developer's preference here by just always exhausting the stream regardless of the flag on the handler?

Reply:
We'll somehow need to document this away and make it clear it is a Handler preference, not a developer preference.
Because if a `BufferedReader` has been used, it has no mechanism to inject buffered data back into the WSGI input stream. So it makes no sense not to exhaust as the data state would be undefined.

I'll remove this redundant exhaust from here.